samples/use-onchip-memory/use-onchip-memory.cpp

```diff
@@ -0,0 +1,142 @@
+/***************************************************************************
+ *
+ *  Copyright Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  Codeplay's ComputeCpp SDK
+ *
+ *  example_vptr.cpp
+ *
+ *  Description:
+ *  Sample code that demonstrates the use of the use_onchip_memory extension
+ *  to SYCL provided by ComputeCpp.
+ *
+ **************************************************************************/
+
+#include <CL/sycl.hpp>
+#include <SYCL/codeplay.hpp>
+#include <iostream>
+
+namespace sycl = cl::sycl;
+namespace access = sycl::access;
+
+// Sample kernel
+// -------------
+// Since kernels are no longer allowed to be declared inline, I dislike
+// declaring a random name in the global namespace that can potentially be
+// repurposed.
+//
+// This kernel performs the same operation as std::iota, but also scales
+// the result by two.
+//
+class scaled_iota {
+ public:
+  using accessor = sycl::accessor<int, 1, access::mode::discard_write,
+                                  access::target::global_buffer>;
+
+  explicit scaled_iota(accessor deviceAccess) noexcept
+      : m_deviceAccess{std::move(deviceAccess)} {}
+
+  void operator()(sycl::nd_item<2> itemId) const noexcept {
+    const auto linearId = itemId.get_global_linear_id();
+    m_deviceAccess[linearId] = linearId * 2;
+  }
+
+ private:
+  accessor m_deviceAccess;
+};
+
+namespace codeplay = sycl::codeplay;
+
+template <class Policy>
+void use_with_policy(Policy policy, sycl::queue& queue) {
+  auto hostData = sycl::vector_class<int>(1024);
+  {
+    auto taskContext = queue.get_context();
+
+    // Notice that the Codeplay property takes a policy argument: this is used
+    // to indicate whether the property is advantageous or genuinely necessary.
+    //
+    auto deviceData = sycl::buffer<int, 1>{
+        hostData.data(),                  //
+        sycl::range<1>(hostData.size()),  //
+        sycl::property_list{
+            sycl::property::buffer::context_bound(taskContext),
+            codeplay::property::buffer::use_onchip_memory(policy)  // <--------
+        }  //                                            ^---------------------
+    };
+
+    deviceData.set_final_data(hostData.data());
```

Comment:
I don't think you need this if you also set hostData.data() as the first argument to the deviceData ctor (or are you thinking of doing a device-only buffer instead?)